NeoML/src/Dnn/DnnSolver.cpp

```diff
@@ -322,57 +297,123 @@ void CDnnSolver::clipGradients(const CObjectArray<CDnnBlob>& paramDiffBlobs)
 	}
 }
 
-static const int DnnSolverVersion = 1;
+static CString concatLayerPath( const CArray<CString>& path )
+{
+	CString layerPath = path[0];
+	for( int i = 1; i < path.Size(); ++i ) {
+		layerPath += layerPathSeparator + path[i];
+	}
+	return layerPath;
+}
+
+void CDnnSolver::loadPrevVersionDnnSolverMaps( CArchive& archive, const CDnn& dnn )
+{
+	CMap<CString, CArray<CString>> layerPrevIdToPath;
+	auto mapLayerIdToPath = [&layerPrevIdToPath]( const CDnnLayerGraph& dnn, auto& mapLayerIdToPath ) -> void
+	{
+		CArray<const char*> layerNames;
+		dnn.GetLayerList( layerNames );
+		for( const char* layerName : layerNames ) {
+			const CBaseLayer* layer = dnn.GetLayer( layerName );
+			const CString layerPath = layer->GetPath( "" );
+			CArray<CString>& path = layerPrevIdToPath.GetOrCreateValue( layerPath );
+			layer->GetPath( path );
+			NeoAssert( path.Size() );
+			const CCompositeLayer* composite = dynamic_cast<const CCompositeLayer*>( layer );
+			if( composite != nullptr ) {
+				mapLayerIdToPath( *composite, mapLayerIdToPath );
+			}
+		}
+	};
+	mapLayerIdToPath( dnn, mapLayerIdToPath );
+
+	auto convertOldIdToLayerPath = [&]( const CBaseLayer** layer )
+	{
+		CString layerId;
+		archive >> layerId;
+		const CArray<CString>& path = layerPrevIdToPath[layerId];
+		if( layer != nullptr ) {
+			*layer = dnn.GetLayer( path );
+		}
+		return concatLayerPath( path );
+	};
+
+	int size;
+	archive >> size;
+	for( int i = 0; i < size; ++i ) {
+		const CBaseLayer* layerTemp = nullptr;
+		const CString layerPath = convertOldIdToLayerPath( &layerTemp );
+
+		CDiffBlobSum& blobSum = layerToParamDiffBlobsSum.GetOrCreateValue( layerPath );
+		archive >> blobSum.Count;
+		SerializeBlobs( mathEngine, archive, blobSum.Sum );
+		blobSum.LayerOwner = layerTemp;
+	}
+
+	archive >> size;
+	for( int i = 0; i < size; ++i ) {
+		const CString layerPath = convertOldIdToLayerPath( nullptr );
+		SerializeBlobs( mathEngine, archive, layerToGradientHistory.GetOrCreateValue( layerPath ) );
+	}
+}
+
+static const int DnnSolverVersion = 2;
 
-void CDnnSolver::Serialize( CArchive& archive, CDnn& dnn )
+void CDnnSolver::Serialize( CArchive& archive, const CDnn& dnn )
 {
 	const int version = archive.SerializeVersion( DnnSolverVersion );
 	if( archive.IsStoring() ) {
-		CMap<CBaseLayer*, CString> layerPtrToId;
-		mapLayerPtrToId( dnn, layerPtrToId );
-
 		archive << layerToParamDiffBlobsSum.Size();
 		for( int pos = layerToParamDiffBlobsSum.GetFirstPosition(); pos != NotFound;
 			pos = layerToParamDiffBlobsSum.GetNextPosition( pos ) )
 		{
-			archive << layerPtrToId[layerToParamDiffBlobsSum.GetKey( pos )];
+			CString layerPath = layerToParamDiffBlobsSum.GetKey( pos );
+			const CBaseLayer* layer = layerToParamDiffBlobsSum.GetValue( pos ).LayerOwner;
+			NeoAssert( layer != nullptr );
+			CArray<CString> path;
+			layer->GetPath( path );
+			archive.Serialize( path );
```

Comment:
тут сериализуется массив. А можно сразу тут хранить строку с разделителями

Reply:
One can concat the string using some delimiter, but one cannot be sure how to split the string back since any delimiter could be part of the layer name.